src/backoff.rs

```diff
@@ -0,0 +1,294 @@
+// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
+
+// https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/
+
+use rand::{thread_rng, Rng};
+use std::time::Duration;
+
+pub trait Backoff: Clone + Send + 'static {
+    // Returns the delay period for next retry. If the maximum retry count is hit returns None.
+    fn next_delay_duration(&mut self) -> Option<Duration>;
+}
+
+#[derive(Clone)]
+pub struct NoJitterBackoff {
+    current_attempts: u32,
+    max_attempts: u32,
+    current_delay_ms: u64,
+    max_delay_ms: u64,
+}
+
+impl NoJitterBackoff {
+    pub const fn new(base_delay_ms: u64, max_delay_ms: u64, max_attempts: u32) -> Self {
+        Self {
+            current_attempts: 0,
+            max_attempts,
+            current_delay_ms: base_delay_ms,
+            max_delay_ms,
+        }
+    }
+}
+
+impl Backoff for NoJitterBackoff {
+    fn next_delay_duration(&mut self) -> Option<Duration> {
+        if self.current_attempts >= self.max_attempts {
+            return None;
+        }
+
+        let delay_ms = self.max_delay_ms.min(self.current_delay_ms);
+
+        self.current_attempts += 1;
+        self.current_delay_ms <<= 1;
+
+        Some(Duration::from_millis(delay_ms))
+    }
+}
+
+#[derive(Clone)]
+pub struct FullJitterBackoff {
```

Comment:
Could you comment on these structs to describe the differences please?

Reply:
I add comment and formula to each struct Backoff, and remove Option from the parameter backoff you mentioned below. Thanks.